wiki/C++-Style-Guide.md

```diff
@@ -0,0 +1,814 @@
+
+Do not create your own classes where you can just use the STL.
+The STL will generally be as fast as anything you can write yourself, and also much less buggy.
+
+
+
+# Functions
+
+This section describes how functions should be named and also lays out some widely-accepted design principles for factoring code into functions.
+
+## Function names
+
+Function names should be CapitalizedCamelCase
+There is no single standard in C++ for whether or not function names are capitalized. But, constructor names must be capitalized, so it makes sense to capitalize the others too. Also, it is easy to distinguish class names from function names or variables based solely on context, but the only easy way to distinguish a function from a function object is by capitalization.
+Exception: function names that match the standard library (such as begin() and end()) may be lower-case.
+
+## Definition order
+
+Functions in the .cpp file should appear in the same order as they are declared in the .h file.
+This makes it easier to scan through the files to find a function, which is often faster than going to the trouble of doing a Find... within the file. It also means that related functions are grouped together instead of appearing in an arbitrary order.
+
+## Const correctness
+
+Any function or function parameter that can be const, should be const.
+It takes a little bit of extra work when defining a class interface, but makes all sorts of subtle errors impossible to commit.
+Note that this means that if a class provides begin() and end() iterators, you need to define both a const and a non-const copy of those functions.
+
+## Exception safety
+
+Use the RAII idiom whenever possible so that code is automatically exception-safe.
+That is, use objects like ofstream, vector, lock_guard, etc. that automatically do whatever cleanup is necessary (closing a file, freeing memory, releasing a mutex) when they are destroyed. That produces cleaner code anyways, and has the advantage that you don't even have to worry about where in the function an exception might get thrown.
+
+## Avoid copy-and-pasted code
+
+If nearly identical code appears in two places, factor it out into a separate function.
+Otherwise, if you make a change or a bug fix, you're likely to make it in one of the copies but not in the other. And, copy-and-paste is often a sign of quickly written throw-away code.
+
+## Non-member functions
+
+All functions should either be a member of a class or of an unnamed namespace.
+This implies that some classes may consist entirely of static functions, and never actually be intended to be instantiated.
+This is so that it is immediately clear to anyone reading the code what file they need to look at to see the declaration or definition of that function. That is, if I see "MyClass::DoStuff(myData);", I know to look at MyClass.h for details. If I see "DoStuff(myData);", I have to read through all the #includes to see which one defined DoStuff().
```

Comment:
```suggestion
This is so that it is immediately clear to anyone reading the code what file they need to look at to see the declaration or definition of that function. That is, if I see `MyClass::DoStuff(myData);`, I know to look at MyClass.h for details. If I see `DoStuff(myData);`, I have to read through all the #includes to see which one defined DoStuff().
```